### DISCUSSION TITLE
Add asStateFlow to the StateFlow variables in ViewModel class

### DISCUSSION THREAD

**Discussion by kaungkhantsoe** [nimblehq/android-templates#518] [RFCs]
## Why
Currently, we define the `StateFlow` variables inside the `ViewModel` as follows:
```
val uiModels: StateFlow<List<UiModel>> = _uiModels
```
One issue with this approach is we can still convert the `StateFlow` variable to a `MutableStateFlow` in the other classes
```
val mutableUiModels = viewModel.uiModels as MutableStateFlow
```
We can ' emit ' new values once we can cast them to `MutableStateFlow` which we should not allow.
```
val coroutineScope = rememberCoroutineScope()
LaunchedEffect(key1 = Unit) {
    coroutineScope.launch {
        mutableUiModels.emit(listOf())
    }
}
```
If we were implementing an SDK, this could be an issue. One way to restrict such conversion would be using `asStateFlow` extension. If we look into the definition of `asStateFlow`, we will see the following code 👇 
```
/**
 * Represents this mutable state flow as a read-only state flow.
 */
public fun <T> MutableStateFlow<T>.asStateFlow(): StateFlow<T> =
    ReadonlyStateFlow(this, null)
```
That way we can restrict the conversion of `StateFlow` to `MutableFlow` by defining the variables in ViewModel like the following:
```
private val _uiModels = MutableStateFlow<List<UiModel>>(emptyList())
val uiModels = _uiModels.asStateFlow()
```

## Who benefits?
Android developers

**Comment by AVI5HEK** [comment]
> Currently, we define the StateFlow variables inside the ViewModel as follows:
> ```kotlin
> val uiModels: StateFlow<List<UiModel>> = _uiModels.asStateFlow()
> ```

You mean, currently we define it like this, right?
```kotlin
val uiModels: StateFlow<List<UiModel>> = _uiModels
```

**Reply by kaungkhantsoe** [reply]
@AVI5HEK Correct. We can even shorten the declaration like this.
```
val uiModels = _uiModels.asStateFlow()
```

**Reply by luongvo** [reply]
> val uiModels = _uiModels.asStateFlow()

This style looks good to me. Voted 👍

**Comment by chornerman** [comment]
Voted 🚀

**Comment by manh-t** [comment]
Agreed. The Kotlin doc also has an example code like this.
![image](https://github.com/nimblehq/android-templates/assets/60863885/36b246a2-96a4-4af1-9d61-b32699605098)

However, we should update the description to be correct as @AVI5HEK mentioned, @kaungkhantsoe .
![image](https://github.com/nimblehq/android-templates/assets/60863885/87a01053-7cb1-4da3-8b4c-5084fb660a62)
We should remove `asStateFlow` here so the other devs aren't confused because the current code doesn't have it.

**Reply by luongvo** [reply]
Exactly, I think we should suggest to change from
```
val uiModels: StateFlow<List<UiModel>> = _uiModels
```
to
```
val uiModels = _uiModels.asStateFlow()
```

**Reply by kaungkhantsoe** [reply]
Updated 🚀 @manh-t @luongvo

**Comment by Wadeewee** [comment]
Agreed and Voted 🚀

**Comment by thiennguyen0196** [comment]
Agreed and voted ✋

**Comment by ryan-conway** [comment]
[Issue created 🙏🏻 ](https://github.com/nimblehq/android-templates/issues/535)
